fil_deal_stats.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"context"
+	"sync"
+)
+
+type dealStats struct {
+	hf                     *heyFil
+	refreshLock            sync.RWMutex
+	dealCountByParticipant map[string]int64
+	latestRefreshErr       error
+}
+
+func (ds *dealStats) start(ctx context.Context) {
+	go func() {
+		for {
+			ds.latestRefreshErr = ds.refresh(ctx)
+			select {
+			case <-ctx.Done():
+				return
+			case <-ds.hf.dealStatsRefreshInterval.C:
+				ds.latestRefreshErr = ds.refresh(ctx)
+			}
+		}
+	}()
+}
+
+func (ds *dealStats) refresh(ctx context.Context) error {
+	ch, err := ds.hf.chainHead(ctx)
+	if err != nil {
+		return err
+	}
+	epoch := ch.Height
+
+	deals, err := ds.hf.stateMarketDeals(ctx)
```

Comment:
are you sure the node this is running against is going to answer this?

in index-observer i got it from the regular snapshot https://github.com/filecoin-shipyard/index-observer/blob/main/main.go#L33 because most nodes( chain.love, glif) don't offer this method

Reply:
The implementation of that method here also uses the s3 snapshot.